nav2_shim_controller/src/shim_controller.cpp

```diff
@@ -0,0 +1,397 @@
+  node->get_parameter(plugin_name_ + ".use_rotate_to_path", use_rotate_to_path_);
+  node->get_parameter(plugin_name_ + ".use_dynamic_threshold", use_dynamic_threshold_);
+  node->get_parameter("controller_frequency", control_frequency);
+
+  transform_tolerance_ = tf2::durationFromSec(transform_tolerance);
+  control_duration_ = 1.0 / control_frequency;
+
+  if (node->get_parameter(plugin_name_ + ".default_plugin", default_plugin_name)) {
+    std::string default_plugin_type = nav2_util::get_plugin_type_param(node, default_plugin_name);
+    default_plugin = plugin_loader_.createSharedInstance(default_plugin_type);
+    RCLCPP_INFO(
+      logger_, "Initializing default plugin \"%s\" with type \"%s\"",
+      default_plugin_name.c_str(), default_plugin_type.c_str());
+
+    try {
+      default_plugin->configure(node, default_plugin_name, tf_, costmap_ros_);
+    } catch (const std::exception & e) {
+      RCLCPP_ERROR(logger_, "Couldn't initialize default plugin! %s", e.what());
+      throw nav2_core::PlannerException("Unable to load default plugin!");
+    }
+    RCLCPP_INFO(logger_, "Initialized default plugin \"%s\"", default_plugin_name.c_str());
+  } else {
+    RCLCPP_ERROR(
+      node->get_logger(), "Can not get 'default_plugin' param value for %s", plugin_name_.c_str());
+  }
+}
+
+void ShimController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type"
+    " shim_controller::ShimController",
+    plugin_name_.c_str());
+
+  if (default_plugin) {default_plugin->cleanup();}
+}
+
+void ShimController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type "
+    "shim_controller::ShimController",
+    plugin_name_.c_str());
+  if (default_plugin) {default_plugin->activate();}
+}
+
+void ShimController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type "
+    "shim_controller::ShimController",
+    plugin_name_.c_str());
+
+  if (default_plugin) {default_plugin->deactivate();}
+}
+
+geometry_msgs::msg::TwistStamped ShimController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed,
+  nav2_core::GoalChecker * goal_checker)
+{
+  goal_checker_ = goal_checker;
+  return calcCmdVel(pose, speed);
+}
+
+geometry_msgs::msg::TwistStamped ShimController::calcCmdVel(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed)
+{
+  // Update for the current goal checker's state
+  geometry_msgs::msg::Pose pose_tolerance;
+  geometry_msgs::msg::Twist vel_tolerance;
+  if (goal_checker_ && goal_checker_->getTolerances(pose_tolerance, vel_tolerance)) {
+    goal_dist_tol_ = pose_tolerance.position.x;
+    goal_yaw_tol_ = tf2::getYaw(pose_tolerance.orientation);
+  } else {
+    RCLCPP_WARN(logger_, "Unable to retrieve goal checker's tolerances!");
+  }
+
+  // Transform path to robot base frame
+  auto transformed_plan = transformGlobalPlan(pose);
```

Comment:
Remove all RPP logic, I don't see why that should be in the file that simply rotates a robot to its current path / goal headings

Reply:
The only time you need to do any of this logic is when a new path comes in. So in the set plan function, you could set a flag that it has changed (or compare the paths to see that there is a change) to then trigger the "check if I even need to rotate to rough heading" logic. If not, just pass through and reset the flag. If so, rotate to heading. Once rotated within the tolerance, reset the flag as pass through to the internal controller plugin.